File: CS-253/Final/fraction.cpp
/* 
 * File:   fraction.cpp
 * Author: krumse01
 * 
 * Created on May 4, 2016, 9:19 AM
 */

#include "fraction.h"

// greatest common divisor code provided
int gcd(int x, int y) {
    if (y==0)
        return x;
    return gcd(y, x%y);
}

Fraction::Fraction() : Fraction(0) {}
// if default constructor is called, just make fraction with numerator=0

Fraction::Fraction(int n) {
    top = n;
    bottom = 1;
    // when only numerator is provided, denominator will always be 1
}

Fraction::Fraction(int n, int d) {
    int div = gcd(n,d);
    top = n/div;
    bottom = d/div;
    // find greatest common divisor, divide numerator and denominator by it
    // to simplify fraction
}

int Fraction::numerator() const {
    return top;
}

int Fraction::denominator() const {
    return bottom;
}

double Fraction::asDouble() const {
    return (double)top/(double)bottom;
}

Fraction Fraction::add(const Fraction& x) const {
    int num = top * x.denominator() + x.numerator() * bottom;
    int den = bottom * x.denominator();
    return Fraction(num,den);
}

Fraction Fraction::subtract(const Fraction& x) const {
    int num = top * x.denominator() - x.numerator() * bottom;
    int den = bottom * x.denominator();
    return Fraction(num,den);
}

Fraction Fraction::divide(const Fraction& x) const {
    int num = top * x.denominator();
    int den = bottom * x.numerator();
    return Fraction(num,den);
}

Fraction Fraction::multiply(const Fraction& x) const {
    int num = top * x.numerator();
    int den = bottom * x.denominator();
    return Fraction(num,den);
}

Fraction operator+(const Fraction& a, const Fraction& b) {
    return a.add(b);
}

Fraction operator-(const Fraction& a, const Fraction& b) {
    return a.subtract(b);
}

Fraction operator*(const Fraction& a, const Fraction& b) {
    return a.multiply(b);
}

Fraction operator/(const Fraction& a, const Fraction& b) {
    return a.divide(b);
}

File: CS-252/Final/Fraction.java

public class Fraction /*extends Object*/ {

    private int top;
    private int bottom;
    private boolean displayDecimal = false;

    public Fraction(int n, int d) {
        int gcd = gcd(n, d);
        top = n / gcd;
        bottom = d / gcd;
    }

    public Fraction(int n) {
        this(n, 1);
    }

    @Override
    public String toString() {
        if (displayDecimal) {
            return ((double) top / bottom) + "";
        } else if (bottom == 1) {
            return top + "";
        } else {
            return top + "/" + bottom;
        }
    }

    public static int gcd(int x, int y) {
        if (x % y == 0) {
            return y;
        }
        return gcd(y, x % y);
    }

    public Fraction add(Fraction other) {
        int top1 = top * other.bottom;
        int top2 = other.top * bottom;
        int bot = bottom * other.bottom;
        return new Fraction(top1 + top2, bot);
    }

    public Fraction subtract(Fraction other) {
        int top1 = top * other.bottom;
        int top2 = other.top * bottom;
        int bot = bottom * other.bottom;
        return new Fraction(top1 - top2, bot);
    }

    public Fraction multiply(Fraction other) {
        return new Fraction(top * other.top, bottom * other.bottom);
    }

    public Fraction divide(Fraction other) {
        return new Fraction(top * other.bottom, bottom * other.top);
    }

    public double decimalForm() {
        return (double) top / bottom;
    }

    public Fraction negate() {
        return new Fraction(top * -1, bottom);
    }

    public void toggleDisplay() {
        displayDecimal = !displayDecimal;
    }
}
